File: python/gamex/_LIB/compression/lzo.py
import os
from io import BytesIO

# Lzo1xDecompressor
class Lzo1xDecompressor:
    def __init__(self, input: bytes, output: bytearray):
        self.input = input
        self.output = output
        self.inputPointer = 0
        self.outputPointer = 0
        self.outputPointer2 = 0

    def read(self) -> int: z = self.input[self.inputPointer]; self.inputPointer += 1; return z
    def peek(self) -> int: return self.input[self.inputPointer]
    def readUInt16(self) -> int: z = self.input[self.inputPointer] + (self.input[self.inputPointer + 1] << 8); self.inputPointer += 2; return z

    def decompress(self) -> int:
        def _eof():
            return 0 if self.inputPointer == len(self.input) else -8 if self.inputPointer < len(self.input) else -4

        gotoFirst = gotoMatchdone = False; t = 0
        _first = True
        if self.peek() > 17:
            t = self.read() - 17
            if t >= 4: gotoFirst = True
            copyBytes(max(1, t))
        while True:
            if gotoFirst: gotoFirst = False #goto _first
            else:
                t = self.read()
                if t >= 16: _first = False #goto _match
                else:
                    if t == 0: t += 15 + self.readLength()
                    self.copyBytes(4 + t - 1)
            
            #_first:
            if _first:
                t = self.read()
                if t >= 16: pass #goto _match
                else:
                    self.outputPointer2 = self.outputPointer - (1 + 0x0800) - (t >> 2) - (self.read() << 2)
                    self.copyBytes(3, fromOutput=True)
                    gotoMatchdone = True
            else: _first = True
            
            #_match:
            while True:
                def f1():
                    nonlocal t
                    if t >= 2 * 4 - (3 - 1) and (self.outputPointer - self.outputPointer2) >= 4: t += 4 - (3 - 1); self.copyBytes(t, fromOutput=True)
                    else: self.copyBytes(max(3, t + 2), fromOutput=True)
                if gotoMatchdone: gotoMatchdone = False # goto _matchDone
                else:
                    if t >= 64:
                        self.outputPointer2 = self.outputPointer - 1 - ((t >> 2) & 7) - (self.read() << 3)
                        t = (t >> 5) - 1
                        self.copyBytes(max(3, t + 2), fromOutput=True)
                        # goto _matchDone
                    elif t >= 32:
                        t &= 31
                        if t == 0: t += 31 + self.readLength()
                        self.outputPointer2 = self.outputPointer - 1 - (self.readUInt16() >> 2)
                        f1()
                    elif t >= 16:
                        self.outputPointer2 = self.outputPointer - ((t & 8) << 11)
                        t &= 7
                        if t == 0: t += 7 + self.readLength()
                        self.outputPointer2 -= self.readUInt16() >> 2
                        if self.outputPointer2 == self.outputPointer: return _eof()
                        self.outputPointer2 -= 0x4000
                        f1()
                    else:
                        self.outputPointer2 = self.outputPointer - 1 - (t >> 2) - (self.read() << 2)
                        self.copyBytes(2, fromOutput=True)
                        # goto _matchDone
                
                # _matchDone:
                t = self.input[self.inputPointer - 2] & 3
                if t == 0: break
                self.copyBytes(t)
                t = self.read()

    def readLength(self) -> int:
        length = 0
        while True:
            b = self.read()
            if b == 0: length += 255
            else: length += b; break
        return length

    def copyBytes(self, size: int, fromOutput: bool = False) -> None:
        if fromOutput:
            for i in range(size): self.output[self.outputPointer + i] = self.output[self.outputPointer2 + i]
            self.outputPointer += size
            self.outputPointer2 += size
        else:
            for i in range(size): self.output[self.outputPointer + i] = self.input[self.inputPointer + i]
            self.outputPointer += size
            self.inputPointer += size

File: dotnet/Core/GameX/_LIB/Compression/Lzo.cs
namespace Compression;

using System;

public class Lzo1xDecompressor(byte[] input, byte[] output) {
    //public static byte[] Decompress(byte[] compressed, int uncompressedLength) {
    //    byte[] decompressed = new byte[uncompressedLength];
    //    var lzoDecompressor = new Lzo1xDecompressor(compressed, decompressed);
    //    var result = lzoDecompressor.Decompress();
    //    if (result != 0) throw new Exception("Decompression failed.");
    //    return decompressed;
    //}
    readonly byte[] Input = input;
    readonly byte[] Output = output;
    uint InputPointer = 0;
    uint OutputPointer = 0;
    uint OutputPointer2;

    byte Read() => Input[InputPointer++];
    byte Peek() => Input[InputPointer];
    ushort ReadUInt16() => (ushort)(Input[InputPointer++] + (Input[InputPointer++] << 8));

    public int Decompress() {
        bool gotoFirst = false, gotoMatchdone = false; uint t;
        if (Peek() > 17) {
            t = (uint)(Read() - 17);
            if (t >= 4) gotoFirst = true;
            CopyBytes(Math.Max(1, t));
        }
        while (true) {
            if (gotoFirst) { gotoFirst = false; goto _first; }
            t = Read();
            if (t >= 16) goto _match;
            if (t == 0) t += 15 + ReadLength();
            CopyBytes(4 + t - 1);
        _first:
            t = Read();
            if (t >= 16) goto _match;
            OutputPointer2 = OutputPointer - (1 + 0x0800) - (t >> 2) - ((uint)Read() << 2);
            CopyBytes(3, fromOutput: true);
            gotoMatchdone = true;
        _match:
            while (true) {
                if (gotoMatchdone) { gotoMatchdone = false; goto _match_done; }
                if (t >= 64) {
                    OutputPointer2 = OutputPointer - 1 - ((t >> 2) & 7) - ((uint)Read() << 3);
                    t = (t >> 5) - 1;
                    CopyBytes(Math.Max(3, t + 2), fromOutput: true);
                    goto _match_done;
                }
                else if (t >= 32) {
                    t &= 31;
                    if (t == 0) t += 31 + ReadLength();
                    OutputPointer2 = OutputPointer - 1 - ((uint)ReadUInt16() >> 2);
                }
                else if (t >= 16) {
                    OutputPointer2 = OutputPointer - ((t & 8) << 11);
                    t &= 7;
                    if (t == 0) t += 7 + ReadLength();
                    OutputPointer2 -= (uint)ReadUInt16() >> 2;
                    if (OutputPointer2 == OutputPointer) goto _eof;
                    OutputPointer2 -= 0x4000;
                }
                else {
                    OutputPointer2 = OutputPointer - 1 - (t >> 2) - ((uint)Read() << 2);
                    CopyBytes(2, fromOutput: true);
                    goto _match_done;
                }
                if (t >= 2 * 4 - (3 - 1) && (OutputPointer - OutputPointer2) >= 4) { t += 4 - (3 - 1); CopyBytes(t, fromOutput: true); }
                else CopyBytes(Math.Max(3, t + 2), fromOutput: true);
            _match_done:
                t = (uint)(Input[InputPointer - 2] & 3);
                if (t == 0) break;
                CopyBytes(t);
                t = Read();
            }
        }
    _eof:
        return InputPointer == Input.Length ? 0 : InputPointer < Input.Length ? -8 : -4;
    }

    uint ReadLength() {
        uint length = 0U;
        while (true) {
            var b = Read();
            if (b == 0) length += 255;
            else { length += b; break; }
        }
        return length;
    }

    void CopyBytes(uint size, bool fromOutput = false) {
        if (fromOutput) {
            for (var i = 0; i < size; i++) Output[OutputPointer + i] = Output[OutputPointer2 + i];
            OutputPointer += size;
            OutputPointer2 += size;
        }
        else {
            for (var i = 0; i < size; i++) Output[OutputPointer + i] = Input[InputPointer + i];
            OutputPointer += size;
            InputPointer += size;
        }
    }

}